File: bindings/kotlin/src/main/kotlin/zag/VersionCheck.kt
package zag

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap

/** CLI version detection and compatibility checking for zag bindings. */
object VersionCheck {

    /** A parsed semver tuple. */
    data class SemVer(val major: Int, val minor: Int, val patch: Int) : Comparable<SemVer> {
        override fun compareTo(other: SemVer): Int {
            major.compareTo(other.major).let { if (it != 0) return it }
            minor.compareTo(other.minor).let { if (it != 0) return it }
            return patch.compareTo(other.patch)
        }
    }

    /** A feature requirement with method name, minimum version, and whether it is set. */
    data class Requirement(val method: String, val version: String, val isSet: Boolean)

    private val versionCache = ConcurrentHashMap<String, String>()

    /** Parse a semver string like "0.6.0" into a numeric tuple. */
    fun parseSemver(version: String): SemVer {
        val parts = version.trim().split(".")
        if (parts.size != 3) {
            throw ZagException(
                "Could not parse version \"$version\": expected format \"X.Y.Z\"",
                null, "")
        }
        try {
            return SemVer(parts[0].toInt(), parts[1].toInt(), parts[2].toInt())
        } catch (e: NumberFormatException) {
            throw ZagException(
                "Could not parse version \"$version\": non-numeric components",
                null, "")
        }
    }

    /**
     * Detect the CLI version by running `{bin} --version`.
     * Result is cached per binary path.
     */
    suspend fun detectVersion(bin: String): String {
        versionCache[bin]?.let { return it }

        return withContext(Dispatchers.IO) {
            val process = try {
                ProcessBuilder(bin, "--version").start()
            } catch (e: Exception) {
                throw ZagException(
                    "Could not detect zag CLI version: failed to run '$bin --version'. " +
                        "Ensure zag is installed and on your PATH, or set ZAG_BIN. (${e.message})",
                    null, "")
            }

            val stdout = process.inputStream.bufferedReader().readText().trim()
            val exitCode = process.waitFor()

            if (exitCode != 0) {
                throw ZagException(
                    "Could not detect zag CLI version: '$bin --version' exited with code $exitCode",
                    exitCode, "")
            }

            val parts = stdout.split("\\s+".toRegex())
            val versionStr = parts.lastOrNull() ?: ""

            // Validate it parses
            parseSemver(versionStr)

            versionCache[bin] = versionStr
            versionStr
        }
    }

    /**
     * Check that the installed CLI version satisfies all configured requirements.
     * Throws [ZagException] if any requirement is not met.
     */
    suspend fun check(bin: String, requirements: List<Requirement>) {
        val active = requirements.filter { it.isSet }
        if (active.isEmpty()) return

        val detected = detectVersion(bin)
        val detectedSv = parseSemver(detected)

        val failures = active.filter { detectedSv < parseSemver(it.version) }
        if (failures.isEmpty()) return

        if (failures.size == 1) {
            val f = failures[0]
            throw ZagException(
                "${f.method} requires zag CLI >= ${f.version}, " +
                    "but the installed version is $detected. Please update the zag binary.",
                null, "")
        }

        val lines = failures.joinToString("\n") { "  - ${it.method} requires >= ${it.version}" }
        throw ZagException(
            "The following methods require a newer zag CLI version:\n$lines\n" +
                "Installed version: $detected. Please update the zag binary.",
            null, "")
    }

    /** Inject a version into the cache for testing. */
    internal fun setVersionForTesting(bin: String, version: String) {
        versionCache[bin] = version
    }

    /** Clear the version cache for testing. */
    internal fun clearVersionCache() {
        versionCache.clear()
    }
}

File: bindings/csharp/src/Zag/VersionCheck.cs
using System.Collections.Concurrent;
using System.Diagnostics;

namespace Zag;

/// <summary>CLI version detection and compatibility checking for zag bindings.</summary>
public static class VersionCheck
{
    /// <summary>Parsed semver tuple.</summary>
    internal readonly record struct SemVer(int Major, int Minor, int Patch) : IComparable<SemVer>
    {
        public int CompareTo(SemVer other)
        {
            int c = Major.CompareTo(other.Major);
            if (c != 0) return c;
            c = Minor.CompareTo(other.Minor);
            if (c != 0) return c;
            return Patch.CompareTo(other.Patch);
        }
    }

    /// <summary>A feature requirement with method name, minimum version, and whether it is set.</summary>
    public readonly record struct Requirement(string Method, string Version, bool IsSet);

    private static readonly ConcurrentDictionary<string, string> VersionCache = new();

    /// <summary>Parse a semver string like "0.6.0" into a numeric tuple.</summary>
    internal static SemVer ParseSemver(string version)
    {
        var parts = version.Trim().Split('.');
        if (parts.Length != 3 ||
            !int.TryParse(parts[0], out var major) ||
            !int.TryParse(parts[1], out var minor) ||
            !int.TryParse(parts[2], out var patch))
        {
            throw new ZagException(
                $"Could not parse version \"{version}\": expected format \"X.Y.Z\"",
                null, "");
        }
        return new SemVer(major, minor, patch);
    }

    /// <summary>Detect the CLI version by running <c>{bin} --version</c>. Cached per binary path.</summary>
    public static async Task<string> DetectVersionAsync(string bin, CancellationToken ct = default)
    {
        if (VersionCache.TryGetValue(bin, out var cached))
            return cached;

        var psi = new ProcessStartInfo(bin, "--version")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        Process process;
        try
        {
            process = Process.Start(psi)
                ?? throw new ZagException(
                    $"Could not detect zag CLI version: failed to start '{bin} --version'.",
                    null, "");
        }
        catch (Exception ex)
        {
            throw new ZagException(
                $"Could not detect zag CLI version: failed to run '{bin} --version'. " +
                $"Ensure zag is installed and on your PATH, or set ZAG_BIN. ({ex.Message})",
                null, "");
        }

        var stdout = await process.StandardOutput.ReadToEndAsync(ct);
        await process.WaitForExitAsync(ct);

        if (process.ExitCode != 0)
        {
            throw new ZagException(
                $"Could not detect zag CLI version: '{bin} --version' exited with code {process.ExitCode}",
                process.ExitCode, "");
        }

        var output = stdout.Trim();
        var parts = output.Split(' ', StringSplitOptions.RemoveEmptyEntries);
        var versionStr = parts.Length > 0 ? parts[^1] : "";

        // Validate it parses
        ParseSemver(versionStr);

        VersionCache[bin] = versionStr;
        return versionStr;
    }

    /// <summary>Check that the installed CLI version satisfies all configured requirements.</summary>
    public static async Task CheckAsync(string bin, IReadOnlyList<Requirement> requirements, CancellationToken ct = default)
    {
        var active = requirements.Where(r => r.IsSet).ToList();
        if (active.Count == 0) return;

        var detected = await DetectVersionAsync(bin, ct);
        var detectedSv = ParseSemver(detected);

        var failures = active
            .Where(r => detectedSv.CompareTo(ParseSemver(r.Version)) < 0)
            .ToList();

        if (failures.Count == 0) return;

        if (failures.Count == 1)
        {
            throw new ZagException(
                $"{failures[0].Method} requires zag CLI >= {failures[0].Version}, " +
                $"but the installed version is {detected}. Please update the zag binary.",
                null, "");
        }

        var lines = string.Join("\n", failures.Select(f => $"  - {f.Method} requires >= {f.Version}"));
        throw new ZagException(
            $"The following methods require a newer zag CLI version:\n{lines}\n" +
            $"Installed version: {detected}. Please update the zag binary.",
            null, "");
    }

    /// <summary>Inject a version into the cache for testing.</summary>
    internal static void SetVersionForTesting(string bin, string version)
    {
        VersionCache[bin] = version;
    }

    /// <summary>Clear the version cache for testing.</summary>
    internal static void ClearVersionCache()
    {
        VersionCache.Clear();
    }
}
